File: modules/uml/app/controllers/uml/UmlController.java
package controllers.uml;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Optional;

import javax.inject.Inject;

import com.fasterxml.jackson.databind.JsonNode;

import controllers.core.ExerciseController;
import model.ClassSelectionResult;
import model.CorrectionException;
import model.DiagramDrawingResult;
import model.UmlExercise;
import model.UmlResult;
import model.UmlSolution;
import model.UmlSolution$;
import model.result.CompleteResult;
import model.user.User;
import play.Logger;
import play.data.DynamicForm;
import play.data.FormFactory;
import play.libs.Json;
import play.mvc.Result;
import play.twirl.api.Html;

public class UmlController extends ExerciseController<UmlExercise, UmlResult> {

  private static final Path SOLUTION_SCHEMA_PATH = Paths.get("conf", "resources", "uml", "solutionSchema.json");

  private static final String ERROR_MSG = "Es gab einen Fehler bei der Validierung des Resultats!";

  public static final JsonNode SOLUTION_SCHEMA_NODE = initSolutionSchemaNode();

  private static final Html EX_LIST_REST = new Html("<div class=\"alert alert-info\">"
      + "Neueinsteiger sollten die Variante mit Zwischenkorrektur verwenden, "
      + "die die einzelnen Schritte der Erstellung eines Klassendiagrammes nach und nach durcharbeitet.</div>\n<hr>");

  @Inject
  public UmlController(FormFactory theFactory) {
    super(theFactory, "uml", UmlExercise.finder, UmlToolObject$.MODULE$);
  }

  private static JsonNode initSolutionSchemaNode() {
    try {
      return Json.parse(String.join("\n", Files.readAllLines(SOLUTION_SCHEMA_PATH)));
    } catch (final IOException e) {
      Logger.error("There has been an error parsing the schema files for UML:", e);
      return null;
    }
  }

  public Result classSelection(int exerciseId) {
    return ok(views.html.classSelection.render(getUser(), finder.byId(exerciseId)));
  }

  public Result correctClassSelection(int exerciseId) {
    try {
      final ClassSelectionResult result = new ClassSelectionResult(finder.byId(exerciseId), readSolutionFromForm());

      return ok(views.html.classSelectionSolution.render(getUser(), result));
    } catch (final CorrectionException e) {
      Logger.error(ERROR_MSG, e);
      return badRequest(e.getMessage());
    }
  }

  public Result correctDiagramDrawing(int exerciseId) {
    try {
      UmlSolution sol = readSolutionFromForm();
      final DiagramDrawingResult result = new DiagramDrawingResult(finder.byId(exerciseId), sol);

      return ok(views.html.diagdrawingsol.render(getUser(), result));

    } catch (final CorrectionException e) {
      Logger.error(ERROR_MSG, e);
      return badRequest(e.getMessage());
    }
  }

  public Result correctDiagramDrawingWithHelp(int exerciseId) {
    try {
      UmlSolution sol = readSolutionFromForm();
      final DiagramDrawingResult result = new DiagramDrawingResult(finder.byId(exerciseId), sol);

      return ok(views.html.diagdrawinghelpsol.render(getUser(), result));
    } catch (final CorrectionException e) {
      Logger.error(ERROR_MSG, e);
      return badRequest(e.getMessage());
    }
  }

  public Result correctMatching(int exerciseId) {
    return ok(views.html.umlMatchingCorrection.render(getUser()));
  }

  public Result diagramDrawing(int exerciseId) {
    return ok(views.html.diagdrawing.render(getUser(), finder.byId(exerciseId), false));
  }

  public Result diagramDrawingWithHelp(int exerciseId) {
    return ok(views.html.diagdrawing.render(getUser(), finder.byId(exerciseId), true));
  }

  public Result matching(int exerciseId) {
    final UmlExercise exercise = finder.byId(exerciseId);
    return ok(views.html.umlMatching.render(getUser(), exercise, exercise.getSolution()));
  }

  private UmlSolution readSolutionFromForm() throws CorrectionException {
    Optional<UmlSolution> sol = UmlSolution$.MODULE$.readFromForm(factory.form().bindFromRequest());

    if(sol.isPresent()) {
      return sol.get();
    } else {
      throw new CorrectionException("", "");
    }
  }

  @Override
  protected CompleteResult<UmlResult> correct(DynamicForm form, UmlExercise exercise, User user) {
    // TODO Auto-generated method stub
    return null;
  }

  @Override
  protected Html renderExercise(User user, UmlExercise exercise) {
    // TODO Auto-generated method stub
    return null;
  }

  @Override
  protected Html renderExesListRest() {
    return EX_LIST_REST;
  }

  @Override
  protected Html renderResult(CompleteResult<UmlResult> correctionResult) {
    // TODO Auto-generated method stub
    return null;
  }

}

File: modules/uml/app/controllers/uml/UmlController.scala
package controllers.uml

import java.nio.file.{Files, Path, Paths}
import javax.inject.Inject

import com.fasterxml.jackson.databind.JsonNode
import controllers.excontrollers.IdPartExController
import model._
import model.result.CompleteResult
import model.user.User
import play.api.data.Form
import play.api.mvc.ControllerComponents
import play.libs.Json
import play.twirl.api.Html

import scala.util.Try

class UmlController @Inject()(cc: ControllerComponents)
  extends IdPartExController[UmlExercise, UmlResult](cc, UmlExercise.finder, UmlToolObject) {

  override type SolType = StringSolution

  override def solForm: Form[StringSolution] = ???

  override def exercise(exerciseId: Int, part: String) = Action { implicit request => {
    val user = getUser
    val exercise = finder.byId(exerciseId)

    Ok(UmlExPart.valueOf(part) match {
      case UmlExPart.CLASS_SELECTION => views.html.classSelection.render(user, exercise)
      case UmlExPart.DIAG_DRAWING => views.html.diagdrawing.render(user, exercise, false)
      case UmlExPart.DIAG_DRAWING_HELP => views.html.diagdrawing.render(user, exercise, true)
      case UmlExPart.ATTRS_METHS => views.html.umlMatching.render(user, exercise)
      case _ => new Html("FEHLER!")
    })
  }
  }

  override def correct(exerciseId: Int, part: String) = Action { implicit request => {
    val exercise = finder.byId(exerciseId)
    Option(null) /*UmlSolution.readFromForm(null factory.form().bindFromRequest())*/ match {
      case None => BadRequest("There has been an error!")
      case Some(sol) =>
        val (result, nextPart) = UmlExPart.valueOf(part) match {
          case UmlExPart.CLASS_SELECTION => (ClassSelectionResult(exercise, sol), UmlExPart.DIAG_DRAWING_HELP)
          case UmlExPart.DIAG_DRAWING_HELP => (DiagramDrawingHelpResult(exercise, sol), UmlExPart.ATTRS_METHS)
          case UmlExPart.DIAG_DRAWING => (DiagramDrawingResult(exercise, sol), UmlExPart.FINISHED)
          case UmlExPart.ATTRS_METHS => (null, UmlExPart.FINISHED)
          case UmlExPart.FINISHED => (null, UmlExPart.FINISHED)
        }
        Ok(views.html.umlResult.render(getUser, result, nextPart))
    }
  }
  }

  override def correctEx(sol: StringSolution, exercise: UmlExercise, user: User): Try[CompleteResult[UmlResult]] = ??? //FIXME: not used...


  override def renderExercise(user: User, exercise: UmlExercise): Html = ??? //FIXME


  override val renderExesListRest = new Html(
    s"""<div class="alert alert-info">
       |Neueinsteiger sollten die Variante mit Zwischenkorrektur verwenden, die die einzelnen Schritte
       |der Erstellung eines Klassendiagrammes nach und nach durcharbeitet.
       |</div>
       |<hr>""")

  override def renderResult(correctionResult: CompleteResult[UmlResult]): Html = ??? //FIXME

}

object UmlController {

  val SolutionSchemaPath: Path = Paths.get("conf", "resources", "uml", "solutionSchema.json")

  val SolutionSchemaNode: JsonNode = Json.parse(String.join("\n", Files.readAllLines(SolutionSchemaPath)))

}
